### PR TITLE
Add support for providing a custom get S2A stream function to the TLSClientConfigFactory

### PR DESCRIPTION
**Pull request by gscert**

### CODE REVIEW

File: s2a_options.go [hunk 1]
```diff
@@ -149,7 +149,7 @@ type ClientOptions struct {
 	FallbackOpts *FallbackOptions
 
 	// Generates an S2AStream interface for talking to the S2A server.
-	getS2AStream func(ctx context.Context, s2av2Address string) (stream.S2AStream, error)
+	GetS2AStream func(ctx context.Context, s2av2Address string) (stream.S2AStream, error)
```
Comment (xmenxk) [inline s2a_options.go:152]: this option is for Google internal use, so we want keep it unexported

Comment (gscert) [reply s2a_options.go]: This has been fixed. Thanks!

File: internal/v2/s2av2_test.go [hunk 2]
```diff
@@ -318,7 +318,18 @@ func (x s2ATestStream) Send(m *s2av2pb.SessionReq) error {
 }
 
 func (x s2ATestStream) Recv() (*s2av2pb.SessionResp, error) {
-	return nil, nil
+	return &s2av2pb.SessionResp{
```
Comment (xmenxk) [inline internal/v2/s2av2_test.go:321]: curious, why is this change needed? were the tests having issue? thx

File: internal/v2/s2av2_e2e_test.go [hunk 3]
```diff
@@ -455,3 +456,24 @@ func TestNewClientTlsConfigWithoutTokenManager(t *testing.T) {
...
+	s2aAddr := startFakeS2A(t, "TestNewClientTlsConfig_token")
+	var tokenManager tokenmanager.AccessTokenManager
+	ctx, cancel := context.WithTimeout(context.Background(), defaultE2ETimeout)
+	defer cancel()
+
+	getStreamFuncCalled := false
+	_, err := NewClientTLSConfig(ctx, s2aAddr, nil, tokenManager, s2av2pb.ValidatePeerCertificateChainReq_CONNECT_TO_GOOGLE, "test_server_name", nil, func(ctx context.Context, s2av2Address string) (stream.S2AStream, error) {
+		getStreamFuncCalled = true
+		return s2ATestStream{debug: "test S2A stream"}, nil
+	})
```
Comment (xmenxk) [inline internal/v2/s2av2_e2e_test.go:471]: This is checking the NewClientTLSConfig() func is called, but IIUC we wanted to test if the passed in getStreamFunc is called?

File: internal/v2/tlsconfigstore/tlsconfigstore.go [hunk 4]
```diff
@@ -75,7 +75,7 @@ func GetTLSConfigurationForClient(serverHostname string, s2AStream stream.S2AStr
 		return nil, fmt.Errorf("failed to get TLS configuration from S2A: %d, %v", resp.GetStatus().Code, resp.GetStatus().Details)
 	}
 
-	// Extract TLS configiguration from SessionResp.
+	// Extract TLS configuration from SessionResp.
```
Comment (xmenxk) [inline internal/v2/tlsconfigstore/tlsconfigstore.go:78]: thank you

File: internal/v2/s2av2_test.go (hunk 2 above)
Comment (gscert) [reply internal/v2/s2av2_test.go:321]: If you follow s2aTLSClientConfigFactory.Build() --> v2.NewClientTLSConfig() --> tlsconfigstore.GetTLSConfigurationForClient, it'll panic on nil ptr dereferences:
  * s2a-go/internal/v2/tlsconfigstore/tlsconfigstore.go:53
  * s2a-go/internal/v2/tlsconfigstore/tlsconfigstore.go:82

File: internal/v2/s2av2_e2e_test.go (hunk 3 above)
Comment (gscert) [reply internal/v2/s2av2_e2e_test.go:471]: It should be testing it underneath: NewClientTLSConfig(..., getStream func param) --> createStream(..., getStream  func param) --> getStream != nil, so invoke.

Can you do a second pass and let me know if you're still seeing this? I might be misunderstanding what you mean.

Comment (xmenxk) [reply internal/v2/s2av2_e2e_test.go:471]: Oh, sry I misread the code, resolving this.

File: internal/v2/s2av2_test.go (hunk 2 above)
Comment (xmenxk) [reply internal/v2/s2av2_test.go:321]: SG

File: internal/v2/s2av2_e2e_test.go [hunk 5]
```diff
@@ -455,3 +456,24 @@ func TestNewClientTlsConfigWithoutTokenManager(t *testing.T) {
 		t.Errorf("tls.Config has unexpected certificate: got: %v, want: %v", got, want)
 	}
 }
+
+func TestNewClientTlsConfigWithCustomS2AStream(t *testing.T) {
+	os.Unsetenv(accessTokenEnvVariable)
+	s2aAddr := startFakeS2A(t, "TestNewClientTlsConfig_token")
+	var tokenManager tokenmanager.AccessTokenManager
+	ctx, cancel := context.WithTimeout(context.Background(), defaultE2ETimeout)
+	defer cancel()
```
Comment (zeromath) [inline internal/v2/s2av2_e2e_test.go:465]: I think it is better to use `t.Cleanup` in unit tests.

https://pkg.go.dev/testing#T.Cleanup

Comment (gscert) [reply internal/v2/s2av2_e2e_test.go]: Done, thanks!
